### PR TITLE
Align gRPC server span status codes to OTEL specs

### LINKED ISSUE #1756: Update gRPC server status code to span status code mapping
Spec ref: https://github.com/open-telemetry/opentelemetry-specification/pull/3333

### PR DESCRIPTION
**Pull request by FilipNikolovski**
# Description

This PR aligns the gRPC server span status codes, to be set based on the changes in the otel specification https://github.com/open-telemetry/opentelemetry-specification/pull/3333.

Fixes #1756

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Updated the `abort` test to check for proper span status against different grpc status codes.

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py [hunk 1]
```diff
@@ -158,25 +154,34 @@ def set_code(self, code):
...
-            )
+            status = self._server_status(self._code, details)
+            self._active_span.set_status(status)
         return self._servicer_context.set_details(details)
 
+    def _server_status(self, code, details):
+        error_status = Status(
+            status_code=StatusCode.ERROR, description=f"{code}:{details}"
+        )
+        status_codes = {
```
Comment (tammy-baylis-swi) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py:172]: Thank you for putting in this PR that lines up with the [spec update](https://github.com/open-telemetry/opentelemetry-specification/pull/3333/files#diff-a4272fe55d12584e896526953d110354d9cc0c4ee26c6cdf559cbafdd50d9cea). The updated tests make sense.

**Comment by tammy-baylis-swi** [comment]
Overall looks good. Please could you resolve the Changelog conflicts as there was a recent new release.

**Comment by tammy-baylis-swi** [comment]
> Thanks! I've resolved the conflicts and the tests, and also updated the Aio grpc server to align the status codes properly with the specs.

Thank you @FilipNikolovski !

Please could you also address the lint suggestions: https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/8189053992/job/22392907101?pr=2019

I see there is also a celery test failure but I don't think it's related to changes in this PR: https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/8189053992/job/22392907101?pr=2019

**Comment by FilipNikolovski** [comment]
> > Thanks! I've resolved the conflicts and the tests, and also updated the Aio grpc server to align the status codes properly with the specs.
> 
> Thank you @FilipNikolovski !
> 
> Please could you also address the lint suggestions: https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/8189053992/job/22392907101?pr=2019
> 
> I see there is also a celery test failure but I don't think it's related to changes in this PR: https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/8189053992/job/22392907101?pr=2019

Took me a while to resolve the failing tests & lint, but anyway..Cleaned up the code a little bit - moved the `_server_status` function to utilities, if you can please have another look :pray:.

**Review by lzchen — APPROVED** [review]
LGTM. It would be great to get this implemented for client as well :)

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+### Fixed
+- Align gRPC span status codes to OTEL specification ([#1756](https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1756))
```
Comment (lzchen) [inline CHANGELOG.md:11]: Oops already merged but this should link to the PR :)

Comment (FilipNikolovski) [reply CHANGELOG.md:11]: ah sorry, should I open up a new PR to fix the link?

**Pull request merged**